src/redisraft.c

```diff
@@ -370,6 +370,161 @@ static void handleInfoCommand(RedisRaftCtx *rr,
     RedisModule_FreeCallReply(reply);
 }
 
+typedef struct MigrationInfo {
+    char shard_group_id[RAFT_DBID_LEN+1];
+    char username[256];
+    char password[256];
+} MigrationInfo;
+
+static MigrationInfo * extractMigrationInfo(RedisRaftCtx *rr, RaftRedisCommand *cmd)
+{
+    MigrationInfo * ret = RedisModule_Calloc(1, sizeof(MigrationInfo));
+    size_t str_len;
+    const char * str = RedisModule_StringPtrLen(cmd->argv[1], &str_len);
+    if (str_len > RAFT_DBID_LEN) {
+        LOG_WARNING("shardgroup id is too long");
+        goto fail;
+    }
+    if (getShardGroupById(rr, str) == NULL) {
+        LOG_WARNING("shardgroup id is invalid: %s", str);
+        goto fail;
+    }
+    strncpy(ret->shard_group_id, str, str_len);
+
+    int idx;
+    for (idx = 6; idx < cmd->argc; idx++) {
+        size_t auth_str_len;
+        const char * auth_str = RedisModule_StringPtrLen(cmd->argv[idx], &auth_str_len);
+        if (auth_str_len == 5 && !strncasecmp("AUTH2", auth_str, 5)) {
+            if (idx + 2 > cmd->argc) {
+                LOG_WARNING("couldn't parse username/password");
+                goto fail;
+            }
+
+            size_t username_len;
+            const char * username = RedisModule_StringPtrLen(cmd->argv[idx+1], &username_len);
+            if (username_len > 255) {
+                LOG_WARNING("username is too long: limited to 255 characters");
+                goto fail;
+            }
+            strncpy(ret->username, username, username_len);
+
+            size_t password_len;
+            const char *password = RedisModule_StringPtrLen(cmd->argv[idx+2], &password_len);
+            if (password_len > 255) {
+                LOG_WARNING("password is too long: limited to 255 characters");
+                goto fail;
+            }
+            strncpy(ret->password, password, password_len);
+            break;
+        } else if (auth_str_len == 4 && !strncasecmp("AUTH", auth_str, 4)) {
+            if (idx + 1 > cmd->argc) {
+                LOG_WARNING("couldn't parse username");
+                goto fail;
+            }
+
+            size_t username_len;
+            const char * username = RedisModule_StringPtrLen(cmd->argv[idx+1], &username_len);
+            if (username_len > 255) {
+                LOG_WARNING("username is too long: limited to 255 characters");
+                goto fail;
+            }
+            strncpy(ret->username, username, username_len);
+            break;
+        }
+    }
+
+    return ret;
+
+fail:
+    RedisModule_Free(ret);
+    return NULL;
+}
+
+static RaftReq * cmdToMigrate(RedisRaftCtx *rr, RedisModuleCtx * ctx, RaftRedisCommand *cmd)
+{
+    RaftReq *req = NULL;
+
+    // 0. assert minimum size here
+    if (cmd->argc < 8) {
+        RedisModule_WrongArity(ctx);
+        return NULL;
+    }
+
+    // 1. Extract migration info and store it in a dict rr->migrate_info
+    MigrationInfo * migrationInfo = extractMigrationInfo(rr, cmd);
```

Comment:
migrationInfo will be leaked in error cases. Btw, as far as I can see, we don't need MigrationInfo at all. Probably you can just parse here and avoid handling another object. Something like:

```c
static RaftReq *cmdToMigrate(RedisRaftCtx *rr, RedisModuleCtx * ctx, RaftRedisCommand *cmd)
{
    RaftReq *req = NULL;
    char username[256] = {0};
    char password[256] = {0};
    char shard_group_id[RAFT_DBID_LEN + 1] = {0};

    // 0. assert minimum size here
    if (cmd->argc < 8) {
        RedisModule_WrongArity(ctx);
        return NULL;
    }

    size_t dbid_len;
    const char *dbid = RedisModule_StringPtrLen(cmd->argv[1], &dbid_len);
    if (dbid_len > RAFT_DBID_LEN) {
        RedisModule_ReplyWithError(ctx, "ERR shardgroup id is too long");
        return NULL;
    }
    strncpy(shard_group_id, dbid, dbid_len);

    for (int idx = 6; idx < cmd->argc; idx++) {
        size_t auth_len;
        const char *auth = RedisModule_StringPtrLen(cmd->argv[idx], &auth_len);
        if (auth_len == 5 && !strncasecmp("AUTH2", auth, 5)) {
            if (idx + 2 > cmd->argc) {
                RedisModule_ReplyWithError(ctx, "ERR couldn't parse username");
                return NULL;
            }

            size_t len;
            const char *ptr;

            ptr = RedisModule_StringPtrLen(cmd->argv[idx + 1], &len);
            if (len > 255) {
                RedisModule_ReplyWithError(ctx, "ERR username is too long");
                return NULL;
            }
            memcpy(username, ptr, len);

            ptr = RedisModule_StringPtrLen(cmd->argv[idx + 2], &len);
            if (len > 255) {
                RedisModule_ReplyWithError(ctx, "ERR password is too long");
                return NULL;
            }
            memcpy(password, ptr, len);
            break;
        } else if (auth_len == 4 && !strncasecmp("AUTH", auth, 4)) {
            if (idx + 1 > cmd->argc) {
                RedisModule_ReplyWithError(ctx, "ERR couldn't parse username");
                return NULL;
            }

            size_t len;
            const char *ptr = RedisModule_StringPtrLen(cmd->argv[idx + 1], &len);
            if (len > 255) {
                RedisModule_ReplyWithError(ctx, "ERR username is too long");
                return NULL;
            }
            memcpy(username, ptr, len);
            break;
        }
    }

    // 2. assert single key is empty
    size_t key_len;
    RedisModule_StringPtrLen(cmd->argv[3], &key_len);
    if (key_len != 0) {
        RedisModule_ReplyWithError(ctx, "ERR RedisRaft doesn't support old style single key migration");
        return NULL;
    }

    /* find "keys" for parsing out keys */
    int idx = 6;
    for(; idx < cmd->argc; idx++) {
        size_t str_len;
        const char * str = RedisModule_StringPtrLen(cmd->argv[idx], &str_len);
        if (str_len == 4 && !strcasecmp("keys", str)) {
            break;
        }
    }

    /* move idx past "keys" argv */
    idx++;

    if (idx >= cmd->argc) {
        RedisModule_ReplyWithError(ctx, "ERR Didn't provide any keys to migrate");
        return NULL;
    }

    RedisModuleString **keys = RedisModule_Alloc(sizeof(RedisModuleString *) * cmd->argc - idx);
    int num_keys = cmd->argc - idx;
    for (int i = 0; i < num_keys; i++) {
        keys[i] = RedisModule_CreateStringFromString(ctx, cmd->argv[idx + i]);
    }

    req = RaftReqInit(ctx, RR_MIGRATE_KEYS);

    /* Overwrite with new data */
    req->r.migrate_keys.num_keys = num_keys;
    req->r.migrate_keys.keys = keys;
    memcpy(req->r.migrate_keys.shardGroupId, shard_group_id, RAFT_DBID_LEN);
    memcpy(req->r.migrate_keys.auth_username, username, 256);
    memcpy(req->r.migrate_keys.auth_password, password, 256);

    return req;
}

```

Reply:
and re migrationInfo leak, yeah, the return NULL was supposed to be a goto exit in my head and just never did that.  done now.